### PR TITLE
[prep CDF-20808, CDF-20999] Robustifying spec against all examples 🪨🐛

### PR DESCRIPTION
**Pull request by doctrino**
# Description

Ran the new ParameterSpec against all `cognite_modules` examples:

* Updated the parameter spec for all the custom implementation done in the loaders.
* Found multiple bugs in our current examples, fixed those...
* ... including a breaking change on Function config for getting consistency on the naming of `dataSetExternalId`.

## Checklist

- [ ] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/_parameters/type_hint.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def _as_str(cls, arg: Any) -> str:
 
     @property
     def types(self) -> list[str]:
-        return [self._as_str(arg) for arg in self.args if not self._is_none_type(arg)]
+        return list({self._as_str(arg) for arg in self.args if not self._is_none_type(arg)})
```
Comment (haakonvt) [inline cognite_toolkit/_cdf_tk/_parameters/type_hint.py:69]: Why not return a set when you ensure uniqueness?

File: cognite_toolkit/_cdf_tk/load/_resource_loaders.py [hunk 2]
```diff
@@ -2132,6 +2263,31 @@ def drop_data(self, ids: SequenceNotStr[NodeId]) -> int:
         # Nodes will be deleted in .delete call.
         return 0
 
+    @classmethod
+    @lru_cache(maxsize=1)
+    def get_write_cls_parameter_spec(cls) -> ParameterSpecSet:
```
Comment (haakonvt) [inline cognite_toolkit/_cdf_tk/load/_resource_loaders.py:2268]: All these `classmethod`s do not seem to use the class.

File: cognite_toolkit/_cdf_tk/load/_resource_loaders.py [hunk 3]
```diff
@@ -2132,6 +2263,31 @@ def drop_data(self, ids: SequenceNotStr[NodeId]) -> int:
...
+        spec = super().get_write_cls_parameter_spec()
+        # Modifications to match the spec
+        for item in spec:
+            if item.path[0] == "apiCall" and len(item.path) > 1:
+                # Move up one level
+                # The spec class is immutable, so we use this trick to modify it.
```
Comment (haakonvt) [inline cognite_toolkit/_cdf_tk/load/_resource_loaders.py:2274]: You can use `dataclasses.replace` I believe, unless you actually need to mutate them directly

File: tests/tests_unit/test_cdf_tk/test_parameters/test_constants.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import pytest
+
+from cognite_toolkit._cdf_tk._parameters import ANY_INT, ANY_STR
+
+
+class TestAnyStrInt:
+    @pytest.mark.parametrize("other", [1, "a", None])
+    def test_any_str_equals(self, other: Any) -> None:
+        is_string = isinstance(other, str)
+        is_equal = ANY_STR == other
```
Comment (haakonvt) [inline tests/tests_unit/test_cdf_tk/test_parameters/test_constants.py:14]: Are these `ANY_` types never instantiated? Perhaps make `__init__` raise?

**Review by haakonvt — APPROVED** [review]
A bit too much magic here for me to fully digest, but the test coverage seems good, so lgtm! 👌

File: cognite_toolkit/_cdf_tk/_parameters/constants.py [hunk 5]
```diff
@@ -1,12 +1,24 @@
+from functools import total_ordering
+
 BASE_TYPES = {t.__name__ for t in (str, int, float, bool)}
 CONTAINER_TYPES = {t.__name__ for t in (list, dict)}
 TYPES = BASE_TYPES | CONTAINER_TYPES
 
 
-class _AnyInt(int):
+@total_ordering
+class _AnyInt(str):
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/_parameters/constants.py:9]: I think that is not valid anymore. Know it is ok.

File: tests/tests_unit/test_cdf_tk/test_parameters/test_constants.py (hunk 4 above)
Comment (doctrino) [reply tests/tests_unit/test_cdf_tk/test_parameters/test_constants.py:14]: Singletons, they are instantiated once.

File: cognite_toolkit/_cdf_tk/load/_resource_loaders.py (hunk 3 above)
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/load/_resource_loaders.py:2274]: I do want to mutate them directly.

File: cognite_toolkit/_cdf_tk/load/_resource_loaders.py (hunk 2 above)
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/load/_resource_loaders.py:2268]: The parent implementation uses the class.

File: cognite_toolkit/_cdf_tk/_parameters/constants.py [hunk 6]
```diff
@@ -1,12 +1,24 @@
...
             return True
+        elif isinstance(other, str):
+            return False
+        return NotImplemented
+
+    def __lt__(self, other: object) -> bool:
+        if isinstance(other, int):
+            return True
+        elif isinstance(other, str):
+            return True
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/_parameters/constants.py:21]: No, it such that AnyInt is always less than any int and any string.

**Pull request merged**